File: archive/canvas-api/src/main/java/com/instructure/canvasapi/api/ModuleAPI.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi.api;

import com.instructure.canvasapi.model.CanvasContext;
import com.instructure.canvasapi.model.MasteryPathSelectResponse;
import com.instructure.canvasapi.model.ModuleItem;
import com.instructure.canvasapi.model.ModuleItemSequence;
import com.instructure.canvasapi.model.ModuleObject;
import com.instructure.canvasapi.utilities.APIHelpers;
import com.instructure.canvasapi.utilities.CanvasCallback;
import com.instructure.canvasapi.utilities.ExhaustiveBridgeCallback;
import com.squareup.okhttp.Response;

import retrofit.Callback;
import retrofit.http.Body;
import retrofit.http.DELETE;
import retrofit.http.GET;
import retrofit.http.POST;
import retrofit.http.PUT;
import retrofit.http.Path;
import retrofit.http.Query;


public class ModuleAPI extends BuildInterfaceAPI {

    public static final String MODULE_ASSET_ASSIGNMENT = "Assignment";
    public static final String MODULE_ASSET_PAGE = "Page";
    public static final String MODULE_ASSET_MODULE_ITEM = "ModuleItem";
    public static final String MODULE_ASSET_QUIZ = "Quiz";
    public static final String MODULE_ASSET_FILE = "File";
    public static final String MODULE_ASSET_DISCUSSION = "Discussion";
    public static final String MODULE_ASSET_EXTERNAL_TOOL = "ExternalTool";

    interface ModulesInterface {
        @GET("/{context_id}/modules")
        void getFirstPageModuleObjects(@Path("context_id") long context_id, Callback<ModuleObject[]> callback);

        @GET("/{next}")
        void getNextPageModuleObjectList(@Path(value = "next", encode = false) String nextURL, Callback<ModuleObject[]> callback);

        @GET("/{context_id}/modules/{module_id}/items?include[]=content_details&include[]=mastery_paths")
        void getFirstPageModuleItems(@Path("context_id") long context_id, @Path("module_id") long moduleID, Callback<ModuleItem[]> callback);

        @GET("/{next}?include[]=content_details&include[]=mastery_paths")
        void getNextPageModuleItemList(@Path(value = "next", encode = false) String nextURL, Callback<ModuleItem[]> callback);

        @POST("/{context_id}/modules/{module_id}/items/{item_id}/mark_read")
        void markModuleItemRead(@Path("context_id") long context_id, @Path("module_id") long module_id, @Path("item_id") long item_id, @Body String body, Callback<Response> callback);

        @PUT("/{context_id}/modules/{module_id}/items/{item_id}/done")
        void markModuleAsDone(@Path("context_id") long context_id, @Path("module_id") long module_id, @Path("item_id") long item_id, @Body String body, Callback<Response> callback);

        @DELETE("/{context_id}/modules/{module_id}/items/{item_id}/done")
        void markModuleAsNotDone(@Path("context_id") long context_id, @Path("module_id") long module_id, @Path("item_id") long item_id, Callback<Response> callback);

        @POST("/{context_id}/modules/{module_id}/items/{item_id}/select_mastery_path")
        void selectMasteryPath(@Path("context_id") long contextId, @Path("module_id") long moduleId, @Path("item_id") long itemId, @Query("assignment_set_id") long assignmentSetId, @Body String body, Callback<MasteryPathSelectResponse> callback);

        @GET("/{context_id}/module_item_sequence")
        void getModuleItemSequence(@Path("context_id") long context_id, @Query("asset_type") String assetType, @Query("asset_id") String assetId, Callback<ModuleItemSequence> callback);
    }

    /////////////////////////////////////////////////////////////////////////
    // API Calls
    /////////////////////////////////////////////////////////////////////////

    public static void getModuleItemsExhaustive(CanvasContext canvasContext, long moduleId, final CanvasCallback<ModuleItem[]> callback) {
        if (APIHelpers.paramIsNull(canvasContext, callback)) { return; }

        CanvasCallback<ModuleItem[]> bridge = new ExhaustiveBridgeCallback<>(ModuleItem.class, callback, new ExhaustiveBridgeCallback.ExhaustiveBridgeEvents() {
            @Override
            public void performApiCallWithExhaustiveCallback(CanvasCallback bridgeCallback, String nextURL, boolean isCached) {
                if(callback.isCancelled()) { return; }

                ModuleAPI.getNextPageModuleItemsChained(nextURL, bridgeCallback, isCached);
            }
        });

        buildCacheInterface(ModulesInterface.class, callback, canvasContext).getFirstPageModuleItems(canvasContext.getId(), moduleId, bridge);
        buildInterface(ModulesInterface.class, callback, canvasContext).getFirstPageModuleItems(canvasContext.getId(), moduleId, bridge);
    }

    public static void getFirstPageModuleObjects(CanvasContext canvasContext, CanvasCallback<ModuleObject[]> callback) {
        if (APIHelpers.paramIsNull(callback, canvasContext)) { return; }

        buildCacheInterface(ModulesInterface.class, callback, canvasContext).getFirstPageModuleObjects(canvasContext.getId(), callback);
        buildInterface(ModulesInterface.class, callback, canvasContext).getFirstPageModuleObjects(canvasContext.getId(), callback);
    }

    public static void getNextPageModuleObjects(String nextURL, CanvasCallback<ModuleObject[]> callback){
        if (APIHelpers.paramIsNull(callback, nextURL)) { return; }

        callback.setIsNextPage(true);
        buildCacheInterface(ModulesInterface.class, callback, false).getNextPageModuleObjectList(nextURL, callback);
        buildInterface(ModulesInterface.class, callback, false).getNextPageModuleObjectList(nextURL, callback);
    }

    public static void getFirstPageModuleItems(CanvasContext canvasContext, long moduleId, CanvasCallback<ModuleItem[]> callback) {
        if (APIHelpers.paramIsNull(callback)) { return; }

        buildCacheInterface(ModulesInterface.class, callback, canvasContext).getFirstPageModuleItems(canvasContext.getId(), moduleId, callback);
        buildInterface(ModulesInterface.class, callback, canvasContext).getFirstPageModuleItems(canvasContext.getId(), moduleId, callback);
    }

    public static void getNextPageModuleItems(String nextURL, CanvasCallback<ModuleItem[]> callback){
        if (APIHelpers.paramIsNull(callback, nextURL)) { return; }

        callback.setIsNextPage(true);
        buildCacheInterface(ModulesInterface.class, callback, false).getNextPageModuleItemList(nextURL, callback);
        buildInterface(ModulesInterface.class, callback, false).getNextPageModuleItemList(nextURL, callback);
    }

    public static void getNextPageModuleItemsChained(String nextURL, CanvasCallback<ModuleItem[]> callback, boolean isCached){
        if (APIHelpers.paramIsNull(callback, nextURL)) { return; }

        callback.setIsNextPage(true);
        if (isCached) {
            buildCacheInterface(ModulesInterface.class, callback, false).getNextPageModuleItemList(nextURL, callback);
        } else {
            buildInterface(ModulesInterface.class, callback, false).getNextPageModuleItemList(nextURL, callback);
        }
    }

    public static void markModuleItemRead(CanvasContext canvasContext, long moduleId, long itemId, CanvasCallback<Response> callback){
        if(APIHelpers.paramIsNull(callback, canvasContext)){
            return;
        }

        buildInterface(ModulesInterface.class, callback, canvasContext).markModuleItemRead(canvasContext.getId(), moduleId, itemId, "", callback);
    }

    public static void markModuleAsDone(CanvasContext canvasContext, long moduleId, long itemId, CanvasCallback<Response> callback){
        if(APIHelpers.paramIsNull(callback, canvasContext)){
            return;
        }

        buildInterface(ModulesInterface.class, callback, canvasContext, false).markModuleAsDone(canvasContext.getId(), moduleId, itemId, "", callback);
    }

    public static void markModuleAsNotDone(CanvasContext canvasContext, long moduleId, long itemId, CanvasCallback<Response> callback){
        if(APIHelpers.paramIsNull(callback, canvasContext)){
            return;
        }

        buildInterface(ModulesInterface.class, callback, canvasContext, false).markModuleAsNotDone(canvasContext.getId(), moduleId, itemId, callback);
    }

    public static void selectMasteryPath(CanvasContext canvasContext, long moduleId, long itemId, long assignmentSetId, CanvasCallback<MasteryPathSelectResponse> callback){
        if(APIHelpers.paramIsNull(callback, canvasContext)){
            return;
        }

        buildInterface(ModulesInterface.class, callback, canvasContext, false).selectMasteryPath(canvasContext.getId(), moduleId, itemId, assignmentSetId, "", callback);
    }

    public static void getModuleItemSequence(CanvasContext canvasContext, String assetType, String assetId, CanvasCallback<ModuleItemSequence> callback) {
        if(APIHelpers.paramIsNull(canvasContext, callback, assetType)) {
            return;
        }

        buildCacheInterface(ModulesInterface.class, callback, canvasContext).getModuleItemSequence(canvasContext.getId(), assetType, assetId, callback);
        buildInterface(ModulesInterface.class, callback, canvasContext).getModuleItemSequence(canvasContext.getId(), assetType, assetId, callback);
    }
}

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/apis/ModuleAPI.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 */

package com.instructure.canvasapi2.apis

import com.instructure.canvasapi2.StatusCallback
import com.instructure.canvasapi2.builders.RestBuilder
import com.instructure.canvasapi2.builders.RestParams
import com.instructure.canvasapi2.models.*
import com.instructure.canvasapi2.models.postmodels.BulkUpdateResponse
import com.instructure.canvasapi2.utils.DataResult
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.http.*

object ModuleAPI {

    interface ModuleInterface {
        @GET("{contextId}/modules")
        fun getFirstPageModuleObjects(@Path("contextId") contextId: Long) : Call<List<ModuleObject>>

        @GET("{contextType}/{contextId}/modules")
        suspend fun getFirstPageModuleObjects(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Tag params: RestParams) : DataResult<List<ModuleObject>>

        @GET
        fun getNextPageModuleObjectList(@Url nextURL: String) : Call<List<ModuleObject>>

        @GET
        suspend fun getNextPageModuleObjectList(@Url nextURL: String, @Tag params: RestParams) : DataResult<List<ModuleObject>>

        @GET("{contextId}/modules?include[]=items&include[]=content_details")
        fun getFirstPageModulesWithItems(@Path("contextId") contextId: Long) : Call<List<ModuleObject>>

        @GET("{contextType}/{contextId}/modules?include[]=items&include[]=content_details")
        suspend fun getFirstPageModulesWithItems(
            @Path("contextType") contextType: String,
            @Path("contextId") contextId: Long,
            @Tag params: RestParams,
            @Query("include[]") includes: List<String> = emptyList()
        ): DataResult<List<ModuleObject>>

        @GET("{contextId}/modules/{moduleId}/items?include[]=content_details&include[]=mastery_paths")
        fun getFirstPageModuleItems(@Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long) : Call<List<ModuleItem>>

        @GET("{contextType}/{contextId}/modules/{moduleId}/items?include[]=content_details&include[]=mastery_paths")
        suspend fun getFirstPageModuleItems(
            @Path("contextType") contextType: String,
            @Path("contextId") contextId: Long,
            @Path("moduleId") moduleId: Long,
            @Tag params: RestParams,
            @Query("include[]") includes: List<String> = emptyList()
        ): DataResult<List<ModuleItem>>

        @GET
        fun getNextPageModuleItemList(@Url nextURL: String) : Call<List<ModuleItem>>

        @GET
        suspend fun getNextPageModuleItemList(@Url nextURL: String, @Tag params: RestParams) : DataResult<List<ModuleItem>>

        @POST("{contextId}/modules/{moduleId}/items/{itemId}/mark_read")
        fun markModuleItemRead(@Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long) : Call<ResponseBody>

        @POST("{contextType}/{contextId}/modules/{moduleId}/items/{itemId}/mark_read")
        suspend fun markModuleItemRead(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long, @Tag params: RestParams) : DataResult<ResponseBody>

        @PUT("{contextType}/{contextId}/modules/{moduleId}/items/{itemId}/done")
        suspend fun markModuleItemAsDone(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long, @Tag params: RestParams) : DataResult<ResponseBody>

        @DELETE("{contextType}/{contextId}/modules/{moduleId}/items/{itemId}/done")
        suspend fun markModuleItemAsNotDone(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long, @Tag params: RestParams): DataResult<ResponseBody>

        @POST("{contextId}/modules/{moduleId}/items/{itemId}/select_mastery_path")
        fun selectMasteryPath(@Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long, @Query("assignment_set_id") assignmentSetId: Long) : Call<MasteryPathSelectResponse>

        @GET("{contextId}/module_item_sequence")
        fun getModuleItemSequence(@Path("contextId") contextId: Long, @Query("asset_type") assetType: String, @Query("asset_id") assetId: String) : Call<ModuleItemSequence>

        @GET("{contextType}/{contextId}/module_item_sequence")
        suspend fun getModuleItemSequence(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Query("asset_type") assetType: String, @Query("asset_id") assetId: String, @Tag params: RestParams) : DataResult<ModuleItemSequence>

        @GET("{contextId}/modules/{moduleId}/items/{itemId}?include[]=content_details")
        fun getModuleItem(@Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long) : Call<ModuleItem>

        @GET("{contextType}/{contextId}/modules/{moduleId}/items/{itemId}?include[]=content_details&include[]=estimated_durations")
        suspend fun getModuleItem(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long, @Tag params: RestParams) : DataResult<ModuleItem>

        @GET("{contextType}/{contextId}/modules/{moduleId}")
        suspend fun getModuleObject(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Tag params: RestParams) : DataResult<ModuleObject>

        @PUT("{contextType}/{contextId}/modules")
        suspend fun bulkUpdateModules(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Query("module_ids[]") moduleIds: List<Long>, @Query("event") event: String, @Query("skip_content_tags") skipContentTags: Boolean, @Query("async") async: Boolean, @Tag params: RestParams): DataResult<BulkUpdateResponse>

        @PUT("{contextType}/{contextId}/modules/{moduleId}/items/{itemId}")
        suspend fun publishModuleItem(@Path("contextType") contextType: String, @Path("contextId") contextId: Long, @Path("moduleId") moduleId: Long, @Path("itemId") itemId: Long, @Query("module_item[published]") publish: Boolean, @Tag params: RestParams): DataResult<ModuleItem>
    }


    fun getAllModuleItems(adapter: RestBuilder, params: RestParams, contextId: Long, moduleId: Long, callback: StatusCallback<List<ModuleItem>>) {
        if (StatusCallback.isFirstPage(callback.linkHeaders)) {
            callback.addCall(adapter.build(ModuleInterface::class.java, params).getFirstPageModuleItems(contextId, moduleId)).enqueue(callback)
        } else {
            callback.addCall(adapter.build(ModuleInterface::class.java, params).getNextPageModuleItemList(callback.linkHeaders?.nextUrl ?: "")).enqueue(callback)
        }
    }

    fun getAllModuleObjects(adapter: RestBuilder, params: RestParams, contextId: Long, callback: StatusCallback<List<ModuleObject>>) {
        if (StatusCallback.isFirstPage(callback.linkHeaders)) {
            callback.addCall(adapter.build(ModuleInterface::class.java, params).getFirstPageModuleObjects(contextId)).enqueue(callback)
        } else {
            callback.addCall(adapter.build(ModuleInterface::class.java, params).getNextPageModuleObjectList(callback.linkHeaders?.nextUrl ?: "")).enqueue(callback)
        }
    }

    fun getFirstPageModuleItems(adapter: RestBuilder, params: RestParams, contextId: Long, moduleId: Long, callback: StatusCallback<List<ModuleItem>>) {
        callback.addCall(adapter.build(ModuleInterface::class.java, params).getFirstPageModuleItems(contextId, moduleId)).enqueue(callback)
    }

    fun getNextPageModuleItems(adapter: RestBuilder, params: RestParams, nextUrl: String, callback: StatusCallback<List<ModuleItem>>) {
        callback.addCall(adapter.build(ModuleInterface::class.java, params).getNextPageModuleItemList(nextUrl)).enqueue(callback)
    }

    fun getFirstPageModuleObjects(adapter: RestBuilder, params: RestParams, contextId: Long, callback: StatusCallback<List<ModuleObject>>) {
        callback.addCall(adapter.build(ModuleInterface::class.java, params).getFirstPageModuleObjects(contextId)).enqueue(callback)
    }

    fun getFirstPageModulesWithItems(
        adapter: RestBuilder,
        params: RestParams,
        contextId: Long,
        callback: StatusCallback<List<ModuleObject>>
    ) {
        callback
            .addCall(adapter.build(ModuleInterface::class.java, params).getFirstPageModulesWithItems(contextId))
            .enqueue(callback)
    }

    fun getNextPageModuleObjects(adapter: RestBuilder, params: RestParams, nextUrl: String, callback: StatusCallback<List<ModuleObject>>) {
        callback.addCall(adapter.build(ModuleInterface::class.java, params).getNextPageModuleObjectList(nextUrl)).enqueue(callback)
    }

    fun markModuleItemAsRead(adapter: RestBuilder, params: RestParams,canvasContext: CanvasContext, moduleId: Long, itemId: Long, callback: StatusCallback<ResponseBody>) {
        callback.addCall(adapter.build(ModuleInterface::class.java, params).markModuleItemRead(canvasContext.id, moduleId, itemId)).enqueue(callback)
    }

    fun selectMasteryPath(adapter: RestBuilder, params: RestParams,canvasContext: CanvasContext, moduleId: Long, itemId: Long, assignmentSetId: Long, callback: StatusCallback<MasteryPathSelectResponse>) {
        callback.addCall(adapter.build(ModuleInterface::class.java, params).selectMasteryPath(canvasContext.id, moduleId, itemId, assignmentSetId)).enqueue(callback)
    }

    fun getModuleItemSequence(adapter: RestBuilder, params: RestParams,canvasContext: CanvasContext, assetType: String, assetId: String, callback: StatusCallback<ModuleItemSequence>) {
        callback.addCall(adapter.build(ModuleInterface::class.java, params).getModuleItemSequence(canvasContext.id, assetType, assetId)).enqueue(callback)
    }

    fun getModuleItem(
        adapter: RestBuilder,
        params: RestParams,
        contextId: Long,
        moduleId: Long,
        moduleItemId: Long,
        callback: StatusCallback<ModuleItem>
    ) {
        callback
            .addCall(
                adapter.build(ModuleInterface::class.java, params).getModuleItem(contextId, moduleId, moduleItemId)
            )
            .enqueue(callback)
    }
}
